### PR TITLE
[MLIR][Linalg] Re-land linalg.matmul move to ODS + Remove/update failing obsolete OpDSL tests for linalg.matmul.

### PR DESCRIPTION
**Pull request by shahidact**
The earlier PR(https://github.com/llvm/llvm-project/pull/104783) which introduces
transpose and broadcast semantic to linalg.matmul was reverted due to two failing
OpDSL test for linalg.matmul.

Since linalg.matmul is now defined using TableGen ODS instead of Python-based OpDSL,
these test started failing and needs to be removed/updated.

This commit removes/updates the failing obsolete tests from below files. All other files
were part of earlier PR and just cherry picked.
    "mlir/test/python/integration/dialects/linalg/opsrun.py"
    "mlir/test/python/integration/dialects/transform.py"

### CODE REVIEW

File: mlir/test/python/integration/dialects/transform.py [hunk 1]
```diff
@@ -99,26 +99,28 @@ def basic(target: any_op_t()):
 # CHECK-LABEL: TEST: test_apply_patterns
 @construct_and_print_in_module
 def test_apply_patterns(module_):
-    M, N, K = 3, 5, 3
+    b, M, N, K = 1, 3, 5, 3
```
Comment (rengolin) [inline mlir/test/python/integration/dialects/transform.py:102]: This is fine for now, but remember, we're do the same with brgemm soon, so we'll need to (re)move this eventually for good.

**Review by rengolin — APPROVED** [review]
Let's give it another go. The fixes here were suggested by the breaking PR, so I'll approve and merge. Thanks!

**Review by MaheshRavishankar — COMMENTED** [review]
I still dont understand whats happening in this change. Its a pretty big change but it is titled as if it is removing obsolete changes. Please describe what is happening with this change and why it is touching all these files. A 900+ line change without accurate title and description is strange.

**Comment by rengolin** [comment]
I have changed the message before merge to reflect that this is a re-land of a previous PR(#104783) with the tests that broke a post-commit bot fixed, as recommended in https://github.com/llvm/llvm-project/pull/104783#issuecomment-2406390905.

**Comment by MaheshRavishankar** [comment]
> I have changed the message before merge to reflect that this is a re-land of a previous PR(#104783) with the tests that broke a post-commit bot fixed, as recommended in [#104783 (comment)](https://github.com/llvm/llvm-project/pull/104783#issuecomment-2406390905).

Ok I see now that this is a reland. I didnt click on the link for the previous PR in the description to see the changes there compared to in this PR and my read of the description/title didnt make that jump as well. Sorry for the noise!

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td [hunk 2]
```diff
@@ -554,6 +554,140 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
...
+     ```
+
+     Example Broadcast and transpose:
+     ```
+     linalg.matmul indexing_maps = [
+                       affine_map<(d0, d1, d2) -> (d2, d0)>, // transpose
+                       affine_map<(d0, d1, d2) -> (d2)>,     // broadcast
+                       affine_map<(d0, d1, d2) -> (d0, d1)>
+                     ]
+                     ins(%arg0, %arg1 : memref<5x3xf32>, memref<7xf32>) outs(%arg2: memref<3x7xf32>)
```
Comment (rolfmorel) [inline mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:605]: Linalg docs broke as this description is missing triple backquotes (```) to close the code block. See https://mlir.llvm.org/docs/Dialects/Linalg/ and search for "Example Broadcast and transpose".

@shahidact , @rengolin

Comment (rengolin) [reply mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:605]: Thanks for fixing it!

**Comment by makslevental** [comment]
FYI after this it's basically impossible to emit a named `linalg.matmul` from python since opdsl pulls double duty. And just restoring the deleted opdsl impl isn't sufficient because the `linalg.MatMulOp` now needs the indexing attributes or else you trip some assert somewhere (ie `idx < size()`). In order to restore `linalg.matmul` I had to

```python
@linalg.linalg_structured_op
def _matmul_generic(
    A=linalg.TensorDef(linalg.T1, linalg.S.M, linalg.S.K),
    B=linalg.TensorDef(linalg.T2, linalg.S.K, linalg.S.N),
    C=linalg.TensorDef(linalg.U, linalg.S.M, linalg.S.N, output=True),
    cast=linalg.TypeFnAttrDef(default=linalg.TypeFn.cast_signed),
):
    linalg.domain(linalg.D.m, linalg.D.n, linalg.D.k)
    linalg.implements(linalg.ContractionOpInterface)
    C[linalg.D.m, linalg.D.n] += cast(linalg.U, A[linalg.D.m, linalg.D.k]) * cast(
        linalg.U, B[linalg.D.k, linalg.D.n]
    )


_matmul_generic.op_name = "matmul"


def matmul(A, B, C, *, loc=None, ip=None):
    if loc is None:
        loc = get_user_code_loc()

    op_configs = linalg.LinalgOpConfig.from_linalg_op_def(
        _matmul_generic.op_def, context=ir.Context.current
    )
    op_config = op_configs[0]
    (
        _all_arg_defs,
        _in_arg_defs,
        _out_arg_defs,
        _outs,
        result_types,
        _type_mapping,
        indexing_maps_attr,
        _iterator_types_attr,
        _index_attrs,
        _fn_attr_mapping,
        _block_arg_types,
    ) = linalg.opdsl.lang.emitter.prepare_common_structured_op(
        op_config.structured_op, A, B, outs=[C], loc=loc, ip=ip
    )
    named_op = linalg.MatmulOp(
        result_types,
        inputs=[A, B],
        outputs=[C],
        indexing_maps=indexing_maps_attr,
        cast=linalg.TypeFn.cast_signed,
        loc=loc,
        ip=ip,
    )
    linalg.fill_builtin_region(named_op.operation)
    return named_op.results
```

I don't know what to propose/suggest here because 

1. I haven't been keeping up with this stream
2. I don't know if anyone even cares - is anyone (except me) using `linalg` from Python?

**Comment by MaheshRavishankar** [comment]
I was afraid of such a thing when the decision was to call the new op added to TableGen the same as the OpDSL name. Previous suggestion was that we change the name of the TableGen version of the op to something else, but that wasnt accepted. I would still suggest doing that to avoid such breakages.

**Comment by rengolin** [comment]
> I was afraid of such a thing when the decision was to call the new op added to TableGen the same as the OpDSL name. Previous suggestion was that we change the name of the TableGen version of the op to something else, but that wasnt accepted. I would still suggest doing that to avoid such breakages.

Having two matmuls will create more confusion than this. We need to move away from OpDSL and the first op is the one that was bound to break things. If we leave the ops in OpDSL, then the Python code will use one set and the C++ code another, and that's way way worse than any temporary code churn.

It seems obvious to me that Python should be able to call other ODS operations. If so, we need to tie the same mechanism to `matmul`, and then the follow up operations. This seems like a mechanical solution would work here.

I don't know the Python code well enough, @makslevental can you help us here?

**Comment by stellaraccident** [comment]
Some mechanical APIs for python would be what is called for here, and the presence or lack of them can't gate the core development. 

Unless if a maintainer stands up who would be willing to pay to update the opdsl infrastructure and separate the concerns, then that path is dead. In the protracted prior debate, no one was willing to make such a commitment, so my strong suspicion is that the only path forward here is for interested parties to develop more mechanical API adaptors/helpers like you have done. I know that's not great, but I'm not sure there is another path open to the project which has costs that anyone will pay.

**Comment by rengolin** [comment]
> It can - there are "binded" builders for every op in every dialect - eg `linalg.MatMulOp` or `linalg.FillOp` etc. The snippet above shows that it can be done (I landed exactly this ["fix" in mlir-python-wheels](https://github.com/makslevental/mlir-python-extras/pull/105)).

If this sounds like the right fix upstream, we should merge it and do the same for the upcoming ops. (@shahidact, @javedabsar, @adam-smnk).

> What makes the current state onerous is that for whatever reason (no doubt due to changes going on in this stream), suddenly you need to explicitly pass the indexing maps even for `linalg.MatMulOp` whereas before you did not. Thus you have to build the `indexing_maps_attr` by [hand](https://github.com/llvm/llvm-project/blob/5dd9867e2d1e698fee980e31da114a37e4c7f612/mlir/python/mlir/dialects/linalg/opdsl/lang/emitter.py#L40-L111) and then call the binded builder.

This was not the intention. The parser accepts plain syntax and builds the affine map by default. We need to find a way for that not to be necessary when the maps are standard. You can keep that logic to support broadcast/transpose maps, though (which is the new thing).

> I highly doubt anyone is using Python to emit `linalg` other than me so I wouldn't call this a catastrophe.

This gives us time to fix it properly.

> I'm not sure there is another path open to the project which has costs that anyone will pay.

If there is, I can't see it either. Neither could anyone else on the original discussion. So let's just all commit to this path and fix all the problems we find along the way.

**Pull request merged**